// gatsby-plugin-google-gtag/dist/gatsby-shared.js
import React from "react";
import { createSnippet } from "./utils/create-snippet";
import { getFlattenedPluginOptions } from "./utils/get-flattened-plugin-options";
import { hasFeature } from "gatsby-plugin-utils";
import { Keys } from "./utils/constants";
export const wrapPageElement = ({
  element
}, pluginOptions) => {
  if (!hasFeature(`gatsby-script`)) {
    return element;
  }

  const {
    Script
  } = require(`gatsby`);

  const flattenedPluginOptions = getFlattenedPluginOptions(pluginOptions);

  const __html = createSnippet(flattenedPluginOptions);

  const {
    gtagURL
  } = flattenedPluginOptions;
  return /*#__PURE__*/React.createElement(React.Fragment, null, element, /*#__PURE__*/React.createElement(Script, {
    key: Keys.gtag,
    src: gtagURL,
    strategy: "off-main-thread",
    forward: [`gtag`]
  }), /*#__PURE__*/React.createElement(Script, {
    key: Keys.gtagConfig,
    id: Keys.gtagConfig,
    strategy: "off-main-thread",
    dangerouslySetInnerHTML: {
      __html
    }
  }));
};
