// gatsby-plugin-google-gtag/dist/gatsby-ssr.js
import React from "react";
import { hasFeature } from "gatsby-plugin-utils";
import { getFlattenedPluginOptions } from "./utils/get-flattened-plugin-options";
import { createSnippet } from "./utils/create-snippet";
import { Keys } from "./utils/constants";
export { wrapPageElement } from "./gatsby-shared";
export const onRenderBody = ({
  setHeadComponents,
  setPostBodyComponents
}, pluginOptions) => {
  const unSupportedEnv = process.env.NODE_ENV !== `production` && process.env.NODE_ENV !== `test`;

  if (unSupportedEnv || hasFeature(`gatsby-script`)) {
    return null;
  }

  const flattenedPluginOptions = getFlattenedPluginOptions(pluginOptions);
  const {
    head,
    gtagURL
  } = flattenedPluginOptions; // Lighthouse recommends pre-connecting to google tag manager

  setHeadComponents([/*#__PURE__*/React.createElement("link", {
    rel: "preconnect",
    key: "preconnect-google-gtag",
    href: origin
  }), /*#__PURE__*/React.createElement("link", {
    rel: "dns-prefetch",
    key: "dns-prefetch-google-gtag",
    href: origin
  })]);
  const setComponents = head ? setHeadComponents : setPostBodyComponents;

  const __html = createSnippet(flattenedPluginOptions);

  return setComponents([/*#__PURE__*/React.createElement("script", {
    key: Keys.gtag,
    async: true,
    src: gtagURL
  }), /*#__PURE__*/React.createElement("script", {
    key: Keys.gtagConfig,
    dangerouslySetInnerHTML: {
      __html
    }
  })]);
};
