// gatsby-plugin-google-gtag/dist/index.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/objectWithoutPropertiesLoose";
const _excluded = ["children"];
import React from "react";
const OutboundLink = /*#__PURE__*/React.forwardRef((_ref, ref) => {
  let {
    children
  } = _ref,
      props = _objectWithoutPropertiesLoose(_ref, _excluded);

  return /*#__PURE__*/React.createElement("a", _extends({
    ref: ref
  }, props, {
    onClick: e => {
      if (typeof props.onClick === `function`) {
        props.onClick(e);
      }

      let redirect = true;

      if (e.button !== 0 || e.altKey || e.ctrlKey || e.metaKey || e.shiftKey || e.defaultPrevented) {
        redirect = false;
      }

      if (props.target && props.target.toLowerCase() !== `_self`) {
        redirect = false;
      } // @ts-ignore gtag is global


      if (window.gtag) {
        // @ts-ignore gtag is global
        window.gtag(`event`, `click`, {
          event_category: `outbound`,
          event_label: props.href,
          transport_type: redirect ? `beacon` : ``,
          // eslint-disable-next-line @typescript-eslint/naming-convention
          event_callback: function () {
            if (redirect) {
              document.location = props.href;
            }
          }
        });
      } else {
        if (redirect) {
          document.location = props.href;
        }
      }

      return false;
    }
  }), children);
});
export { OutboundLink };
